### DISCUSSION TITLE
Question on computing a material vector

### DISCUSSION THREAD

**Discussion by xueyang94** [idaholab/moose#19152] [Q&A Modules: General] [answered]
Hello, 

I am writing a material class, which can compute a vector of material properties, but I couldn't get the initialization right. I have tried `_mat_name(declareProperty<std::vector<MaterialPropertyName>>("mat_names"))`, `_mat_prop(_some_num)`,  and then use a `for` loop to declare for example `_mat_prop[i] = &declareProperty<Real>(_mat_name[i])`.   The complier error is `no matching constructor for initialization of 'std::vector<MaterialPropertyName>'`. (The parent class is `DerivativeMaterialInterface<Material>`.)

Is it possible at all for a material class to compute a material vector? If so, what would be the proper initialization and declaration syntax? I did not find an example material that does this. Thanks.

**Comment by GiudGiud** [comment, marked as answer]
It's possible. Please have a look at the GenericConstantMaterial, which defines a vector of Real material properties. 
https://github.com/idaholab/moose/blob/next/framework/src/materials/GenericConstantMaterial.C

I think for mat_name you want to initialize the vector of names with a getParam<std::vector < MaterialPropertyName > >
Then in the constructor body you want to make the calls to declare each property. 
I dont think declareProperty can be called with a vector of names like that, but I could be wrong

**Reply by GiudGiud** [reply]
If you need a material property vector instead of a vector of material properties, see the answer below

**Reply by xueyang94** [reply]
Update:
Since the vector of materials is computed in the material class, the _prop_values in GenericConstantMaterial cannot be pre-defined in the input file. However, based on the idea of declaring both the prop_name and prop_value, it is fairly easy to output a vector of material properties like this, in case someone needs it:

`params.addRequiredParam<std::vector<MaterialPropertyName>>("some_names", "a description");`
Initializer:
`_some_names(getParam<std::vector<MaterialPropertyName>>("some_names")),`
`_some_prop(_num),`
Declare prop:
`for (unsigned int i=0; i<_num; ++i)`
 `_some_prop[i] = &declareProperty<Real>(_some_names[i]);`

**Reply by GiudGiud** [reply]
Agreed, the treatment of values will have to be different. I m pointing to this material for the handling of the vectors.

**Comment by ngrilli** [comment]
Dear @xueyang94 

This is an example from my repository:
https://github.com/ngrilli/c_pfor_am/blob/main/src/materials/FiniteStrainCrystalPlasticityBackstress.C

For instance in the constructor, I declare:

_rho_for(declareProperty<std::vector<Real>>("rho_for")),

Then I provide the dimension of the vector:

_rho_for[_qp].resize(_nss);

before starting using it.
Then I initialise:

  for (unsigned int i = 0; i < _nss; ++i) // initialise forest dislocation densities
    _rho_for[_qp][i] = _init_rho_for;

Trust this helps,
Best Regards,
Nicolò

**Reply by xueyang94** [reply]
Thanks for your reply. In your repo, `_rho_for` is declared as a vector of 'Real'. However, I want to declare a vector of material properties, which can be output by name and taken by other materials or kernels in the input file. For example, in the input file, I want to have `some_mat = "mat1, mat2, mat3,.."` as one of my material class's parameters. This is similar but not exactly the same as what your code does. I appreciate your advice.

Xueyang

**Reply by GiudGiud** [reply]
so you want to build a vector material property from a vector of material properties?
Like mat1, mat2, mat3 are already defined, and you want to access them by just passing some_mat = "mat1 mat2 mat3" to a kernel

**Reply by xueyang94** [reply]
@GiudGiud Not really. mat1, mat2, and mat3 are not already known. They should be computed by myMatClass, where they are the elements of a material vector. In the input file that uses myMatClass, it looks like:
`[something]`
`type = myMatClass`
`some_mat = "mat1, mat2, mat3"`
`....`
`[../]`

This way, mat1, mat2, and mat3 can be used in the kernels. The diffusion kernel can take mat1 as the diffusivity:
`[Diffusion]`
`type = MatDiffusion`
`variable = u`
`diffusivity = mat1`
`[../]`

mat2 and mat3 can also be used somewhere else in the kernels.

**Reply by GiudGiud** [reply]
This feels like my first post is appropriate then.
GenericConstantMaterial, GenericFunctionMaterial etc all do something like that. 
The syntax for GenericConstantMaterial would be:
```
[material_name]
  type = GenericConstantMaterial
  prop_names = "mat1 mat2 mat3"
  prop_values = "1 2 43"
[]
```

**Reply by xueyang94** [reply]
@GiudGiud Yup, I was about the try the method in your post, but then I found another bug in my code. I am working on fixing that first. I will come back to this if I need more help. Thanks Guillaume!
